engine/table/src/main/java/io/deephaven/engine/util/OuterJoinTools.java

```diff
@@ -0,0 +1,410 @@
+        int numAttempts = 0;
+        String sentinelColumnName;
+        final Map<String, ? extends ColumnSource<?>> resultColumns = leftTable.getColumnSourceMap();
+        do {
+            sentinelColumnName = "__sentinel_" + (numAttempts++) + "__";
+        } while (resultColumns.containsKey(sentinelColumnName));
+
+        // handle the simpler scenario where no conflicts may exist otherwise leftOuterJoin would have failed
+        if (columnsToAdd.length == 0) {
+            final MatchPair[] matchPairs = Arrays.stream(columnsToMatch)
+                    .map(mp -> new MatchPair(mp.rightColumn(), mp.leftColumn()))
+                    .toArray(MatchPair[]::new);
+
+            final Table leftWithSentinel = table1.coalesce().updateView(sentinelColumnName + " = true");
+
+            // we must use leftOuterJoin again as there may be multiple matches per group
+            final Table rightTable = CrossJoinHelper.leftOuterJoin(
+                    (QueryTable) table2.coalesce(),
+                    (QueryTable) leftWithSentinel,
+                    matchPairs,
+                    createColumnsToAdd(leftWithSentinel, matchPairs, MatchPair.ZERO_LENGTH_MATCH_PAIR_ARRAY),
+                    CrossJoinHelper.DEFAULT_NUM_RIGHT_BITS_TO_RESERVE).where(sentinelColumnName + " == null")
+                    .dropColumns(sentinelColumnName);
+
+            return TableTools.merge(leftTable, rightTable);
+        }
+
+        // only need match columns from the left; rename and drop remaining to avoid column name conflicts
+        final SelectColumn[] leftColumns = Streams.concat(
+                Arrays.stream(columnsToMatch).map(mp -> new SourceColumn(mp.leftColumn(), mp.rightColumn())),
+                List.of(SelectColumn.of(Selectable.parse(sentinelColumnName + " = true"))).stream())
+                .toArray(SelectColumn[]::new);
+
+        // we will modify the join to use exact matches
+        final MatchPair[] matchColumns = Arrays.stream(columnsToMatch)
+                .map(mp -> new MatchPair(mp.rightColumn(), mp.rightColumn()))
+                .toArray(MatchPair[]::new);
+
+        // prepare filter for columnsToAdd
+        final SelectColumn[] rightColumns = Arrays.stream(columnsToAdd)
+                .map(mp -> new SourceColumn(mp.rightColumn(), mp.leftColumn()))
+                .toArray(SelectColumn[]::new);
+
+        // we must use leftOuterJoin again as there may be multiple matches per group
+        final Table rightOnlyTable = CrossJoinHelper.leftOuterJoin(
+                (QueryTable) table2.coalesce(),
+                (QueryTable) table1.coalesce().view(leftColumns),
+                matchColumns,
+                new MatchPair[] {new MatchPair(sentinelColumnName, sentinelColumnName)},
+                CrossJoinHelper.DEFAULT_NUM_RIGHT_BITS_TO_RESERVE).where(sentinelColumnName + " == null")
+                .view(rightColumns);
+
+        final Table nullLeftTable = TableTools.newTable(
```

Comment:
I am skeptical that it makes sense to have another join as part of line 123 instead of creating the columns correctly in the view on line 118.